src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,426 @@
+                {
+                    CompletedHandle = completedHandle;
+                    TimedOut = timedOut;
+                }
+
+                public RegisteredWaitHandle CompletedHandle { get; }
+                public bool TimedOut { get; }
+            }
+
+            /// <summary>
+            /// The wait handles registered on this wait thread.
+            /// </summary>
+            private readonly RegisteredWaitHandle[] _registeredWaits = new RegisteredWaitHandle[WaitHandle.MaxWaitHandles - 1];
+            /// <summary>
+            /// The raw wait handles to wait on.
+            /// </summary>
+            /// <remarks>
+            /// The zeroth element of this array is always <see cref="_changeHandlesEvent"/>.
+            /// </remarks>
+            private readonly WaitHandle[] _waitHandles = new WaitHandle[WaitHandle.MaxWaitHandles];
+            /// <summary>
+            /// The number of user-registered waits on this wait thread.
+            /// </summary>
+            private int _numUserWaits = 0;
+            /// <summary>
+            /// A lock for editing any handle registration (i.e. the fields above).
+            /// </summary>
+            private readonly LowLevelLock _registeredHandlesLock = new LowLevelLock();
+
+            /// <summary>
+            /// A list of removals of wait handles that are waiting for the wait thread to process.
+            /// </summary>
+            private readonly RegisteredWaitHandle[] _pendingRemoves = new RegisteredWaitHandle[WaitHandle.MaxWaitHandles - 1];
+            /// <summary>
+            /// The number of pending removals.
+            /// </summary>
+            private int _numPendingRemoves = 0;
+            /// <summary>
+            /// A lock for modifying the pending removals.
+            /// </summary>
+            private readonly LowLevelLock _removesLock = new LowLevelLock();
+
+            /// <summary>
+            /// An event to notify the wait thread that there are pending adds or removals of wait handles so it needs to wake up.
+            /// </summary>
+            private readonly AutoResetEvent _changeHandlesEvent = new AutoResetEvent(false);
+
+            internal bool AnyUserWaits => _numUserWaits != 0;
+
+            public WaitThread()
+            {
+                _waitHandles[0] = _changeHandlesEvent;
+                RuntimeThread waitThread = RuntimeThread.Create(WaitThreadStart);
+                waitThread.IsBackground = true;
+                waitThread.Start();
+            }
+
+            /// <summary>
+            /// The main routine for the wait thread.
+            /// </summary>
+            private void WaitThreadStart()
+            {
+                while (true)
+                {
+                    ProcessRemovals();
+                    int numUserWaits = _numUserWaits;
+                    int preWaitTimeMs = Environment.TickCount;
```

Comment:
Upper bounds checking has been added to the `uint` and `long` overloads in `ThreadPool.cs`.